src/main/flight/pid_init.c

```diff
@@ -66,6 +68,70 @@ static void pidSetTargetLooptime(uint32_t pidLooptime)
 #endif
 }
 
+
+#ifdef USE_WING
+void tpaSpeedBasicInit(const pidProfile_t *pidProfile)
+{
+    // basic model assumes prop pitch speed is inf
+    const float gravityFactor = pidProfile->tpa_speed_basic_gravity / 100.0f;
+    const float delaySec = pidProfile->tpa_speed_basic_delay / 1000.0f;
+
+    pidRuntime.tpaSpeed.twr = 1.0f / (gravityFactor * gravityFactor);
+    const float massDragRatio = (2.0f / logf(3.0f)) * (2.0f / logf(3.0f)) * pidRuntime.tpaSpeed.twr * G_ACCELERATION * delaySec * delaySec;
+    pidRuntime.tpaSpeed.dragMassRatio = 1.0f / massDragRatio;
+    pidRuntime.tpaSpeed.maxSpeed = sqrtf(massDragRatio * pidRuntime.tpaSpeed.twr * G_ACCELERATION + G_ACCELERATION);
+    pidRuntime.tpaSpeed.inversePropMaxSpeed = 0.0f;
+}
+
+
+void tpaSpeedAdvancedInit(const pidProfile_t *pidProfile)
+{
+    // Advanced model uses prop pitch speed, and is quite limited when craft speed is far above prop pitch speed.
+    pidRuntime.tpaSpeed.twr = (float)pidProfile->tpa_speed_adv_thrust / (float)pidProfile->tpa_speed_adv_mass;
+    const float mass = pidProfile->tpa_speed_adv_mass / 1000.0f;
+    const float dragK = pidProfile->tpa_speed_adv_drag_k / 10000.0f;
+    const float propPitch = pidProfile->tpa_speed_adv_prop_pitch / 100.0f;
+    pidRuntime.tpaSpeed.dragMassRatio = dragK / mass;
+    const float propMaxSpeed = (2.54f / 100.0f / 60.0f) * propPitch * motorConfig()->kv * pidRuntime.tpaSpeed.maxVoltage;
+    if (propMaxSpeed <= 0.0f) { // assuming propMaxSpeed is inf
+        pidRuntime.tpaSpeed.inversePropMaxSpeed = 0.0f;
+    } else {
+        pidRuntime.tpaSpeed.inversePropMaxSpeed = 1.0f / propMaxSpeed;
+    }
+
+    const float maxFallSpeed = sqrtf(mass * G_ACCELERATION / dragK);
+
+    const float a = dragK;
+    const float b = mass * pidRuntime.tpaSpeed.twr * G_ACCELERATION * pidRuntime.tpaSpeed.inversePropMaxSpeed;
+    const float c = -mass * (pidRuntime.tpaSpeed.twr + 1) * G_ACCELERATION;
+
+    const float maxDiveSpeed = (-b + sqrtf(b*b - 4.0f * a * c)) / (2.0f * a);
+
+    pidRuntime.tpaSpeed.maxSpeed = MAX(maxFallSpeed, maxDiveSpeed);
+    UNUSED(pidProfile);
+}
+
+
+void tpaSpeedInit(const pidProfile_t *pidProfile)
+{
+    pidRuntime.tpaSpeed.speed = 0.0f;
+    pidRuntime.tpaSpeed.maxVoltage = pidProfile->tpa_speed_max_voltage / 100.0f;
+    pidRuntime.tpaSpeed.pitchOffset = pidProfile->tpa_speed_pitch_offset * M_PIf / 10.0f / 180.0f;
+
+    switch (pidProfile->tpa_speed_type) {
+    case TPA_SPEED_BASIC:
+        tpaSpeedBasicInit(pidProfile);
+        break;
+    case TPA_SPEED_ADVANCED:
+        tpaSpeedAdvancedInit(pidProfile);
+        break;
+    default:
+        break;
+    }
+}
+#endif // USE_WING
+
+
```

Comment:
extra line
```suggestion
```

Reply:
i thought we encourage two lines between function definitions?

Reply:
Not I'm aware off. And if so it is not being used consistently :stuck_out_tongue: 

Can you point out where this was encouraged?

Reply:
i guess it was only in my imagination. fixed :)